x/concentrated-liquidity/incentives.go

```diff
@@ -49,3 +51,161 @@ func (k Keeper) getUptimeAccumulators(ctx sdk.Context, poolId uint64) ([]accum.A
+		// Get relevant uptime-level values
+		curUptimeDuration := types.SupportedUptimes[uptimeIndex]
+
+		// Qualifying liquidity is the amount of liquidity that satisfies uptime requirements
+		qualifyingLiquidity, err := uptimeAccum.GetTotalShares()
+		if err != nil {
+			return err
+		}
+
+		// If there is no qualifying liquidity for the current uptime accumulator, we leave it unchanged
+		if qualifyingLiquidity.LT(sdk.OneDec()) {
+			continue
+		}
+
+		incentivesToAddToCurAccum, updatedPoolRecords, err := calcAccruedIncentivesForAccum(ctx, curUptimeDuration, qualifyingLiquidity, timeElapsedSec, poolIncentiveRecords)
+		if err != nil {
+			return err
+		}
+
+		// Emit incentives to current uptime accumulator
+		uptimeAccum.AddToAccumulator(incentivesToAddToCurAccum)
+
+		// Update pool records (stored in state after loop)
+		poolIncentiveRecords = updatedPoolRecords
+	}
+
+	// Update pool incentive records and LastLiquidityUpdate time in state to reflect emitted incentives
+	pool.SetPoolIncentives(poolIncentiveRecords)
+	pool.SetLastLiquidityUpdate(ctx.BlockTime())
+	err = k.setPool(ctx, pool)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// nolint: unused
+// calcAccruedIncentivesForAccum calculates IncentivesPerLiquidity to be added to an accum
+// Returns the IncentivesPerLiquidity value and an updated list of IncentiveRecords that
+// reflect emitted incentives
+// Returns error if the qualifying liquidity/time elapsed are zero.
+func calcAccruedIncentivesForAccum(ctx sdk.Context, accumUptime time.Duration, qualifyingLiquidity sdk.Dec, timeElapsed sdk.Dec, poolIncentiveRecords []types.IncentiveRecord) (sdk.DecCoins, []types.IncentiveRecord, error) {
+	if !qualifyingLiquidity.IsPositive() || !timeElapsed.IsPositive() {
+		return sdk.DecCoins{}, []types.IncentiveRecord{}, fmt.Errorf("Qualifying liquidity and time elapsed must both be positive.")
+	}
+
+	incentivesToAddToCurAccum := sdk.NewDecCoins()
+	for incentiveIndex, incentiveRecord := range poolIncentiveRecords {
+		// We consider all incentives matching the current uptime that began emitting before the current blocktime
+		if incentiveRecord.StartTime.UTC().Before(ctx.BlockTime().UTC()) && incentiveRecord.MinUptime == accumUptime {
+			// Total amount emitted = time elapsed * emission
+			totalEmittedAmount := timeElapsed.Mul(incentiveRecord.EmissionRate)
+
+			// Incentives to emit per unit of qualifying liquidity = total emitted / qualifying liquidity
+			// Note that we truncate to ensure we do not overdistribute incentives
+			incentivesPerLiquidity := totalEmittedAmount.Quo(qualifyingLiquidity)
+			emittedIncentivesPerLiquidity := sdk.NewDecCoinFromDec(incentiveRecord.IncentiveDenom, incentivesPerLiquidity)
+
+			// Ensure that we only emit if there are enough incentives remaining to be emitted
+			remainingRewards := poolIncentiveRecords[incentiveIndex].RemainingAmount
+			if totalEmittedAmount.LTE(remainingRewards) {
```

Comment:
Well this would only not hold if a specific incentive runs out of tokens, in which case we probably don't want to be erroring for any tx that would trigger a liquidity update (i.e. the caller of these helpers)

Reply:
With this setup if incentives run out it just stops emitting – only problem I can see is if incentives build up over a long period without an update and then none get distributed because they all "run out", although this fuzzy cutoff is difficult to avoid with accumulator stuff (can probably improve in some way though – open to suggestions)